internal/backend/runtime/omni/state_validation.go

```diff
@@ -421,8 +424,83 @@ func machineSetValidationOptions(st state.State, etcdBackupStoreFactory store.Fa
 }
 
 // machineClassValidationOptions returns the validation options for the machine class resource.
+//
+//nolint:gocognit
 func machineClassValidationOptions(st state.State) []validated.StateOption {
+	validateSchema := func(autoProvision *specs.MachineClassSpec_Provision, providerStatus *infra.ProviderStatus) error {
+		if providerStatus.TypedSpec().Value.Schema == "" {
+			return nil
+		}
+
+		filename := fmt.Sprintf("%s-schema.json", providerStatus.Metadata().ID())
+
+		compiler := jsonschema.NewCompiler()
+		if err := compiler.AddResource(filename, strings.NewReader(providerStatus.TypedSpec().Value.Schema)); err != nil {
+			return fmt.Errorf("failed to load json schema for provider %q: %w", autoProvision.ProviderId, err)
+		}
+
+		schema, err := compiler.Compile(filename)
+		if err != nil {
+			return fmt.Errorf("failed to load json schema for provider %q: %w", autoProvision.ProviderId, err)
+		}
+
+		var v interface{}
+		if err = yaml.Unmarshal([]byte(autoProvision.ProviderData), &v); err != nil {
+			return fmt.Errorf("failed to unmarshal provider data %w", err)
+		}
+
+		if err = schema.Validate(v); err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	validate := func(ctx context.Context, res *omni.MachineClass) error {
+		if res.TypedSpec().Value.AutoProvision != nil && res.TypedSpec().Value.MatchLabels != nil {
+			return errors.New("can't set both auto provision and match labels at the same time")
+		}
+
+		if res.TypedSpec().Value.AutoProvision != nil {
+			autoProvision := res.TypedSpec().Value.AutoProvision
+
+			if autoProvision.ProviderId == "" {
+				return errors.New("providerID can not be empty")
+			}
+
+			if autoProvision.TalosVersion == "" {
+				return errors.New("talos version can not be empty")
+			}
+
+			if err := validateTalosVersion(ctx, st, "", autoProvision.TalosVersion); err != nil {
+				return err
+			}
+
+			providerStatus, err := safe.ReaderGetByID[*infra.ProviderStatus](ctx, st, autoProvision.ProviderId)
+			if err != nil {
+				return fmt.Errorf("failed to get provider: %w", err)
+			}
+
+			if err = validateSchema(autoProvision, providerStatus); err != nil {
+				return err
+			}
+		}
+
+		_, err := labels.ParseSelectors(res.TypedSpec().Value.MatchLabels)
```

Comment:
do we need to check for `nil` here ? or it doesn't care?

Reply:
well, maybe we can check that it shouldn't be empty